htlcswitch/link.go

```diff
@@ -2404,24 +2404,17 @@ func (l *channelLink) CheckHtlcForward(payHash [32]byte,
 	policy := l.cfg.FwrdingPolicy
 	l.RUnlock()
 
-	// First check whether the outgoing htlc satisfies the channel policy.
-	err := l.canSendHtlc(
-		policy, payHash, amtToForward, outgoingTimeout, heightNow,
-	)
-	if err != nil {
-		return err
-	}
-
-	// Next, using the amount of the incoming HTLC, we'll calculate the
-	// expected fee this incoming HTLC must carry in order to satisfy the
-	// constraints of the outgoing link.
+	// Using the amount of the incoming HTLC, we'll calculate the expected
+	// fee this incoming HTLC must carry in order to satisfy the constraints
+	// of the outgoing link.
 	expectedFee := ExpectedFee(policy, amtToForward)
 
 	// If the actual fee is less than our expected fee, then we'll reject
 	// this HTLC as it didn't provide a sufficient amount of fees, or the
 	// values have been tampered with, or the send used incorrect/dated
 	// information to construct the forwarding information for this hop. In
-	// any case, we'll cancel this HTLC.
+	// any case, we'll cancel this HTLC. We're checking for this case first
+	// to prevent any information from leaking through 0-fee channel probing.
```

Comment:
For my own understanding, "0-fee probing" would be when you send a probe with no fees through and then get information about the channel based on the error (rather than it first being failed because the fees are wrong)?

I agree with moving the fee check earlier to leak less information, but why wouldn't the person probing just set the correct fees? Since they're not going to pay them anyway if it's a probe?

Reply:
You're right. If the sender is, for example, using keysend with an invalid pre-image, then this PR doesn't actually mitigate any security risks. This is more of a cleanup. But like you said, I think in general it's best to leak as little information as possible.

Reply:
def def, I'd just update the comment here to be more generic? We're introducing a new concept of "0-fee probing" to future readers, when we could rather just give a general explanation that we want to leak as little as possible.